memh5: can't serialize numpy arrays in history

`MemGroup.to_hdf5` fails if there's a numpy array in the history, because it can't get JSON serialized.
This affects ch_util unit test `test_andata.py::TestLoadSave.test_load_FO`:

```
ch_util/tests/test_andata.py ..........F

==================================================================================================== FAILURES =====================================================================================================
____________________________________________________________________________________________ TestLoadSave.test_load_FO ____________________________________________________________________________________________

self = <ch_util.tests.test_andata.TestLoadSave testMethod=test_load_FO>, args = (), kwargs = {}

    def do_test(self, *args, **kwargs):
        with warnings.catch_warnings():
            warnings.simplefilter("ignore", category=UserWarning)
>           test_func(self, *args, **kwargs)

ch_util/tests/test_andata.py:36: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ch_util/tests/test_andata.py:223: in setUp
    data.save(self.test_fname, mode="w")
../pipeline/caput/caput/memh5.py:1828: in save
    **kwargs
../pipeline/caput/caput/memh5.py:545: in to_hdf5
    convert_dataset_strings=convert_dataset_strings,
../pipeline/caput/caput/memh5.py:2276: in deep_group_copy
    convert_attribute_strings=convert_attribute_strings,
../pipeline/caput/caput/memh5.py:2264: in deep_group_copy
    copyattrs(g1.attrs, g2.attrs, convert_strings=convert_attribute_strings)
../pipeline/caput/caput/memh5.py:2222: in copyattrs
    val = _map_json(val)
../pipeline/caput/caput/memh5.py:2214: in _map_json
    value = json_prefix + encoder.encode(value)
/usr/lib/python3.7/json/encoder.py:199: in encode
    chunks = self.iterencode(o, _one_shot=True)
/usr/lib/python3.7/json/encoder.py:257: in iterencode
    return _iterencode(o, 0)
../pipeline/caput/caput/memh5.py:2208: in default
    return json.JSONEncoder.default(self, v)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <caput.memh5.copyattrs.<locals>._map_json.<locals>.DatetimeJSONEncoder object at 0x7f0a70de54d0>, o = array([1.25e+08])

    def default(self, o):
        """Implement this method in a subclass such that it returns
        a serializable object for ``o``, or calls the base implementation
        (to raise a ``TypeError``).
    
        For example, to support arbitrary iterators, you could
        implement default like this::
    
            def default(self, o):
                try:
                    iterable = iter(o)
                except TypeError:
                    pass
                else:
                    return list(iterable)
                # Let the base class default method raise the TypeError
                return JSONEncoder.default(self, o)
    
        """
>       raise TypeError(f'Object of type {o.__class__.__name__} '
                        f'is not JSON serializable')
E       TypeError: Object of type ndarray is not JSON serializable

/usr/lib/python3.7/json/encoder.py:179: TypeError
```

`history["acq"]` was
```
{'CTRL_clock_frequency': array([1.25e+08]), 'FPGA_board_frequency': array([2.e+08]), 'acq.base_path': b'/data/', 'acq.fpga_hk.channel.t_adc_ambient': b'0', 'acq.fpga_hk.channel.t_adc_chip1': b'1', 'acq.fpga_hk.channel.t_adc_chip2': b'2', 'acq.frames_per_file': array([360], dtype=int32), 'acq.len_frame_buf': array([2], dtype=int32), 'acq.serial.path': b'/dev/ttyUSB0', 'acq.serial.timeout': array([0.5]), 'acq.udp.buf_len': array([5], dtype=int32), 'acq.udp.max_len': array([32768], dtype=int32), 'acq.udp.port': array([41001], dtype=int32), 'acq.udp.spf': array([40], dtype=int32), 'adc_board_adc_chip_temperature': array([62.3125, 64.75  ]), 'adc_board_temperature': array([30.03125]), 'adc_clocks': array([2.e+08, 2.e+08, 2.e+08, 2.e+08, 2.e+08, 2.e+08, 2.e+08, 2.e+08]), 'ant_clock': array([2.e+08]), 'antenna_adc_data_acquisition_delay_tables': array([13, 10,  9, 10,  9, 10,  9,  9,  7,  7,  7,  7,  7,  7,  7,  7, 11,
       11,  8,  9,  7,  8,  8,  7,  6,  6,  6,  6,  6,  6,  6,  6, 14, 14,
       14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 13, 13, 13,
       13, 13, 13, 13, 13,  0,  0,  0,  0,  0,  0,  0,  0], dtype=int32), 'antenna_data_source': array([b'adc', b'adc', b'adc', b'adc', b'adc', b'adc', b'adc', b'adc'],
      dtype='|S4'), 'antenna_fft_bypass': array([0, 0, 0, 0, 0, 0, 0, 0], dtype=int32), 'antenna_fft_shift_schedule': array([31, 31, 31, 31, 31, 31, 31, 31], dtype=int32), 'antenna_scaler_log2_gain': array([1, 1, 1, 1, 1, 1, 1, 1], dtype=int32), 'chan_indices': array([(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7),
       (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 2),
       (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 3), (3, 4), (3, 5),
       (3, 6), (3, 7), (4, 4), (4, 5), (4, 6), (4, 7), (5, 5), (5, 6),
       (5, 7), (6, 6), (6, 7), (7, 7)],
      dtype=[('ant_chan_a', '<u2'), ('ant_chan_b', '<u2')]), 'config_capture_time': array([1.37419081e+09]), 'config_protocol_version': array([1, 0], dtype=int32), 'correlator_capture_period_in_frames': array([390625, 390625, 390625, 390625, 390625, 390625, 390625, 390625],
      dtype=int32), 'correlator_clock': array([2.5e+08]), 'correlator_integration_period_in_frames': array([390625, 390625, 390625, 390625, 390625, 390625, 390625, 390625],
      dtype=int32), 'fmc_ref_clock': array([10000000.]), 'fpga.adc_delay.ch00': array([b'13', b'10', b'9', b'10', b'9', b'10', b'9', b'9'], dtype='|S3'), 'fpga.adc_delay.ch01': array([b'7', b'7', b'7', b'7', b'7', b'7', b'7', b'7'], dtype='|S2'), 'fpga.adc_delay.ch02': array([b'11', b'11', b'8', b'9', b'7', b'8', b'8', b'7'], dtype='|S3'), 'fpga.adc_delay.ch03': array([b'6', b'6', b'6', b'6', b'6', b'6', b'6', b'6'], dtype='|S2'), 'fpga.adc_delay.ch04': array([b'14', b'14', b'14', b'14', b'14', b'14', b'14', b'14'],
      dtype='|S3'), 'fpga.adc_delay.ch05': array([b'14', b'14', b'14', b'14', b'14', b'14', b'14', b'14'],
      dtype='|S3'), 'fpga.adc_delay.ch06': array([b'13', b'13', b'13', b'13', b'13', b'13', b'13', b'13'],
      dtype='|S3'), 'fpga.adc_delay.ch07': array([b'0', b'0', b'0', b'0', b'0', b'0', b'0', b'0'], dtype='|S2'), 'fpga.fft_shift': array([31], dtype=int32), 'fpga.int_period': array([0.25]), 'fpga.ip_address': b'10.10.10.11', 'fpga.log2_gain': array([1], dtype=int32), 'fpga.port': array([41000], dtype=int32), 'fpga.ref_freq': array([10000000.]), 'fpga.samp_freq': array([800.]), 'hdf5_version': array([2.]), 'mgt_ref_clock': array([0.]), 'mgt_word_clock': array([0.]), 'n_antenna': array([8], dtype=int32), 'n_freq': array([1024], dtype=int32), 'number_of_antennas': array([8], dtype=int32), 'number_of_antennas_to_correlate': array([8], dtype=int32), 'number_of_correlators': array([8], dtype=int32), 'number_of_correlators_max': array([8], dtype=int32), 'run_message': b'restarting after reloading firmware due to unexpected chFPGA crash- ma', 'system_firmware_version': b'2013-05-02 19:39:41', 'system_frame_length': array([2048], dtype=int32), 'system_frame_period': array([2.56e-06]), 'system_list_of_antennas_with_channelizers': array([0, 1, 2, 3, 4, 5, 6, 7], dtype=int32), 'system_list_of_implemented_correlators': array([0, 1, 2, 3, 4, 5, 6, 7], dtype=int32), 'system_platform_id': array([1], dtype=int32), 'system_reference_frequency': array([10000000.]), 'system_sampling_frequency': array([8.e+08]), 'system_user': b'chime', 'archive_version': '1.0.0'}
```

--- a/caput/memh5.py
+++ b/caput/memh5.py
@@ -2197,18 +2197,32 @@ def copyattrs(a1, a2, convert_strings=False):
     def _map_json(value):
         # Serialize/deserialize "special" json values
 
-        # Datetimes often appear in the configs (as they are parsed by PyYAML),
-        # so we need to serialise them back to strings
-        class DatetimeJSONEncoder(json.JSONEncoder):
+        class Memh5JSONEncoder(json.JSONEncoder):
+            """
+            - Datetimes often appear in the configs (as they are parsed by PyYAML),
+              so we need to serialise them back to strings.
+            - Some old data format may have numpy arrays in `history["acq"]`. We have to convert
+              those to lists and decode byte objects.
+            """
+
             def default(self, v):
                 if isinstance(v, datetime.datetime):
                     return v.isoformat()
+                elif isinstance(v, np.number):
+                    return v.data
+                elif isinstance(v, np.ndarray):
+                    if len(v) == 1:
+                        return v.tolist()[0]
+                    return v.tolist()
+                elif isinstance(v, bytes):  # pragma: py3
+                    return v.decode()
+
                 # Let the default method raise the TypeError
                 return json.JSONEncoder.default(self, v)
 
         if isinstance(value, dict) and isinstance(a2, h5py.AttributeManager):
             # Save to JSON converting datetimes.
-            encoder = DatetimeJSONEncoder()
+            encoder = Memh5JSONEncoder()
             value = json_prefix + encoder.encode(value)
         elif isinstance(value, str) and value.startswith(json_prefix):
             # Read from JSON, keep serialised datetimes as strings

